activerecord/lib/active_record/migration/command_recorder.rb

```diff
@@ -318,6 +318,13 @@ def invert_remove_exclusion_constraint(args)
           super
         end
 
+        def invert_add_unique_key(args)
+          options = args.dup.extract_options!
+
+          raise ActiveRecord::IrreversibleMigration, "add_unique_key is not reversible if given an using_index." if options[:using_index]
```

Comment:
Would you tell me the reason why?

Reply:
When deleting a constraint, the index on which it depends is also deleted.
This means that it is not possible to delete a constraint and rollback only its index.

Reply:
Understood.

I've found that `alter table <tablename> add constraint <constraintname> unique using index <indexname>` renames
the unique index name to the constraint name, therefore when the constraint is dropped, it also drops the renamed unique index.

```sql
test=# ALTER TABLE "sections" ADD CONSTRAINT "uniq_rails_79b901ffb4" UNIQUE USING INDEX "unique_index";
NOTICE:  ALTER TABLE / ADD CONSTRAINT USING INDEX will rename index "unique_index" to "uniq_rails_79b901ffb4"
```

- SQL statements to test
```sql
CREATE UNLOGGED TABLE "sections" ("id" bigserial primary key, "position" integer NOT NULL);
\d "sections"
CREATE UNIQUE INDEX "unique_index" ON "sections" ("position");
\d "sections"
ALTER TABLE "sections" ADD CONSTRAINT "uniq_rails_79b901ffb4" UNIQUE USING INDEX "unique_index";
\d "sections"
ALTER TABLE "sections" DROP CONSTRAINT "uniq_rails_79b901ffb4";
\d "sections"
```

- Result

```sql
test=# select version();
-[ RECORD 1 ]------------------------------------------------------------------------------------------------------------------------
version | PostgreSQL 14.7 (Ubuntu 14.7-1.pgdg22.10+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 12.2.0-3ubuntu1) 12.2.0, 64-bit
test=# CREATE UNLOGGED TABLE "sections" ("id" bigserial primary key, "position" integer NOT NULL);
CREATE TABLE
test=# \d "sections"
                         Unlogged table "public.sections"
  Column  |  Type   | Collation | Nullable |               Default
----------+---------+-----------+----------+--------------------------------------
 id       | bigint  |           | not null | nextval('sections_id_seq'::regclass)
 position | integer |           | not null |
Indexes:
    "sections_pkey" PRIMARY KEY, btree (id)

test=# CREATE UNIQUE INDEX "unique_index" ON "sections" ("position");
CREATE INDEX
test=# \d "sections"
                         Unlogged table "public.sections"
  Column  |  Type   | Collation | Nullable |               Default
----------+---------+-----------+----------+--------------------------------------
 id       | bigint  |           | not null | nextval('sections_id_seq'::regclass)
 position | integer |           | not null |
Indexes:
    "sections_pkey" PRIMARY KEY, btree (id)
    "unique_index" UNIQUE, btree ("position")

test=# ALTER TABLE "sections" ADD CONSTRAINT "uniq_rails_79b901ffb4" UNIQUE USING INDEX "unique_index";
NOTICE:  ALTER TABLE / ADD CONSTRAINT USING INDEX will rename index "unique_index" to "uniq_rails_79b901ffb4"
ALTER TABLE
test=# \d "sections"
                         Unlogged table "public.sections"
  Column  |  Type   | Collation | Nullable |               Default
----------+---------+-----------+----------+--------------------------------------
 id       | bigint  |           | not null | nextval('sections_id_seq'::regclass)
 position | integer |           | not null |
Indexes:
    "sections_pkey" PRIMARY KEY, btree (id)
    "uniq_rails_79b901ffb4" UNIQUE CONSTRAINT, btree ("position")

test=# ALTER TABLE "sections" DROP CONSTRAINT "uniq_rails_79b901ffb4";
ALTER TABLE
test=# \d "sections"
                         Unlogged table "public.sections"
  Column  |  Type   | Collation | Nullable |               Default
----------+---------+-----------+----------+--------------------------------------
 id       | bigint  |           | not null | nextval('sections_id_seq'::regclass)
 position | integer |           | not null |
Indexes:
    "sections_pkey" PRIMARY KEY, btree (id)

test=#
```

Reply:
https://github.com/postgres/postgres/commit/88452d5ba6b adds this feature that is available since PostgreSQL 9.1.